### ISSUE TITLE
[bug] sts get caller identity does not work immediately

### ISSUE THREAD

**Issue by hjkatz** [aws/aws-sdk-go-v2#2517]
### Describe the bug

See: https://github.com/aws/aws-sdk-go-v2/discussions/2093

### Expected Behavior

I expect the call to `sts.GetCallerIdentity()` to succeed immediately after authenticating/receiving new sts credentials.

### Current Behavior

The calls to `sts.GetCallerIdentity()` seem to be inconsistently failing.

### Reproduction Steps

See: https://github.com/aws/aws-sdk-go-v2/discussions/2093

### Possible Solution

See: https://github.com/aws/aws-sdk-go-v2/discussions/2093

### AWS Go SDK V2 Module Versions Used

github.com/aws/aws-sdk-go@v1.44.52 github.com/jmespath/go-jmespath@v0.4.0
github.com/aws/aws-sdk-go@v1.44.52 github.com/pkg/errors@v0.9.1
github.com/aws/aws-sdk-go@v1.44.52 golang.org/x/net@v0.0.0-20220127200216-cd36cc0744dd
github.com/aws/aws-sdk-go-v2@v1.24.1 github.com/aws/smithy-go@v1.19.0
github.com/aws/aws-sdk-go-v2@v1.24.1 github.com/google/go-cmp@v0.5.8
github.com/aws/aws-sdk-go-v2@v1.24.1 github.com/jmespath/go-jmespath@v0.4.0
github.com/aws/aws-sdk-go-v2/config@v1.26.6 github.com/aws/aws-sdk-go-v2@v1.24.1
github.com/aws/aws-sdk-go-v2/config@v1.26.6 github.com/aws/aws-sdk-go-v2/credentials@v1.16.16
github.com/aws/aws-sdk-go-v2/config@v1.26.6 github.com/aws/aws-sdk-go-v2/feature/ec2/imds@v1.14.11
github.com/aws/aws-sdk-go-v2/config@v1.26.6 github.com/aws/aws-sdk-go-v2/internal/ini@v1.7.3
github.com/aws/aws-sdk-go-v2/config@v1.26.6 github.com/aws/aws-sdk-go-v2/service/sso@v1.18.7
github.com/aws/aws-sdk-go-v2/config@v1.26.6 github.com/aws/aws-sdk-go-v2/service/ssooidc@v1.21.7
github.com/aws/aws-sdk-go-v2/config@v1.26.6 github.com/aws/aws-sdk-go-v2/service/sts@v1.26.7
github.com/aws/aws-sdk-go-v2/config@v1.26.6 github.com/aws/smithy-go@v1.19.0
github.com/aws/aws-sdk-go-v2/config@v1.26.6 github.com/google/go-cmp@v0.5.8
github.com/aws/aws-sdk-go-v2/config@v1.26.6 github.com/aws/aws-sdk-go-v2/internal/configsources@v1.2.10
github.com/aws/aws-sdk-go-v2/config@v1.26.6 github.com/aws/aws-sdk-go-v2/internal/endpoints/v2@v2.5.10
github.com/aws/aws-sdk-go-v2/config@v1.26.6 github.com/aws/aws-sdk-go-v2/service/internal/accept-encoding@v1.10.4
github.com/aws/aws-sdk-go-v2/config@v1.26.6 github.com/aws/aws-sdk-go-v2/service/internal/presigned-url@v1.10.10
github.com/aws/aws-sdk-go-v2/credentials@v1.16.16 github.com/aws/aws-sdk-go-v2@v1.24.1
github.com/aws/aws-sdk-go-v2/credentials@v1.16.16 github.com/aws/aws-sdk-go-v2/feature/ec2/imds@v1.14.11
github.com/aws/aws-sdk-go-v2/credentials@v1.16.16 github.com/aws/aws-sdk-go-v2/service/sso@v1.18.7
github.com/aws/aws-sdk-go-v2/credentials@v1.16.16 github.com/aws/aws-sdk-go-v2/service/ssooidc@v1.21.7
github.com/aws/aws-sdk-go-v2/credentials@v1.16.16 github.com/aws/aws-sdk-go-v2/service/sts@v1.26.7
github.com/aws/aws-sdk-go-v2/credentials@v1.16.16 github.com/aws/smithy-go@v1.19.0
github.com/aws/aws-sdk-go-v2/credentials@v1.16.16 github.com/google/go-cmp@v0.5.8
github.com/aws/aws-sdk-go-v2/credentials@v1.16.16 github.com/aws/aws-sdk-go-v2/internal/configsources@v1.2.10
github.com/aws/aws-sdk-go-v2/credentials@v1.16.16 github.com/aws/aws-sdk-go-v2/internal/endpoints/v2@v2.5.10
github.com/aws/aws-sdk-go-v2/credentials@v1.16.16 github.com/aws/aws-sdk-go-v2/service/internal/accept-encoding@v1.10.4
github.com/aws/aws-sdk-go-v2/credentials@v1.16.16 github.com/aws/aws-sdk-go-v2/service/internal/presigned-url@v1.10.10
github.com/aws/aws-sdk-go-v2/feature/s3/manager@v1.15.15 github.com/aws/aws-sdk-go-v2@v1.24.1
github.com/aws/aws-sdk-go-v2/feature/s3/manager@v1.15.15 github.com/aws/aws-sdk-go-v2/config@v1.26.6
github.com/aws/aws-sdk-go-v2/feature/s3/manager@v1.15.15 github.com/aws/aws-sdk-go-v2/service/s3@v1.48.1
github.com/aws/aws-sdk-go-v2/feature/s3/manager@v1.15.15 github.com/aws/smithy-go@v1.19.0
github.com/aws/aws-sdk-go-v2/feature/s3/manager@v1.15.15 github.com/google/go-cmp@v0.5.8
github.com/aws/aws-sdk-go-v2/feature/s3/manager@v1.15.15 github.com/aws/aws-sdk-go-v2/aws/protocol/eventstream@v1.5.4
github.com/aws/aws-sdk-go-v2/feature/s3/manager@v1.15.15 github.com/aws/aws-sdk-go-v2/credentials@v1.16.16
github.com/aws/aws-sdk-go-v2/feature/s3/manager@v1.15.15 github.com/aws/aws-sdk-go-v2/feature/ec2/imds@v1.14.11
github.com/aws/aws-sdk-go-v2/feature/s3/manager@v1.15.15 github.com/aws/aws-sdk-go-v2/internal/configsources@v1.2.10
github.com/aws/aws-sdk-go-v2/feature/s3/manager@v1.15.15 github.com/aws/aws-sdk-go-v2/internal/endpoints/v2@v2.5.10
github.com/aws/aws-sdk-go-v2/feature/s3/manager@v1.15.15 github.com/aws/aws-sdk-go-v2/internal/ini@v1.7.3
github.com/aws/aws-sdk-go-v2/feature/s3/manager@v1.15.15 github.com/aws/aws-sdk-go-v2/internal/v4a@v1.2.10
github.com/aws/aws-sdk-go-v2/feature/s3/manager@v1.15.15 github.com/aws/aws-sdk-go-v2/service/internal/accept-encoding@v1.10.4
github.com/aws/aws-sdk-go-v2/feature/s3/manager@v1.15.15 github.com/aws/aws-sdk-go-v2/service/internal/checksum@v1.2.10
github.com/aws/aws-sdk-go-v2/feature/s3/manager@v1.15.15 github.com/aws/aws-sdk-go-v2/service/internal/presigned-url@v1.10.10
github.com/aws/aws-sdk-go-v2/feature/s3/manager@v1.15.15 github.com/aws/aws-sdk-go-v2/service/internal/s3shared@v1.16.10
github.com/aws/aws-sdk-go-v2/feature/s3/manager@v1.15.15 github.com/aws/aws-sdk-go-v2/service/sso@v1.18.7
github.com/aws/aws-sdk-go-v2/feature/s3/manager@v1.15.15 github.com/aws/aws-sdk-go-v2/service/ssooidc@v1.21.7
github.com/aws/aws-sdk-go-v2/feature/s3/manager@v1.15.15 github.com/aws/aws-sdk-go-v2/service/sts@v1.26.7
github.com/aws/aws-sdk-go-v2/feature/s3/manager@v1.15.15 github.com/jmespath/go-jmespath@v0.4.0
github.com/aws/aws-sdk-go-v2/service/cloudwatchlogs@v1.31.0 github.com/aws/aws-sdk-go-v2@v1.24.1
github.com/aws/aws-sdk-go-v2/service/cloudwatchlogs@v1.31.0 github.com/aws/aws-sdk-go-v2/aws/protocol/eventstream@v1.5.4
github.com/aws/aws-sdk-go-v2/service/cloudwatchlogs@v1.31.0 github.com/aws/aws-sdk-go-v2/internal/configsources@v1.2.10
github.com/aws/aws-sdk-go-v2/service/cloudwatchlogs@v1.31.0 github.com/aws/aws-sdk-go-v2/internal/endpoints/v2@v2.5.10
github.com/aws/aws-sdk-go-v2/service/cloudwatchlogs@v1.31.0 github.com/aws/smithy-go@v1.19.0
github.com/aws/aws-sdk-go-v2/service/cloudwatchlogs@v1.31.0 github.com/google/go-cmp@v0.5.8
github.com/aws/aws-sdk-go-v2/service/ec2@v1.146.0 github.com/aws/aws-sdk-go-v2@v1.24.1
github.com/aws/aws-sdk-go-v2/service/ec2@v1.146.0 github.com/aws/aws-sdk-go-v2/internal/configsources@v1.2.10
github.com/aws/aws-sdk-go-v2/service/ec2@v1.146.0 github.com/aws/aws-sdk-go-v2/internal/endpoints/v2@v2.5.10
github.com/aws/aws-sdk-go-v2/service/ec2@v1.146.0 github.com/aws/aws-sdk-go-v2/service/internal/accept-encoding@v1.10.4
github.com/aws/aws-sdk-go-v2/service/ec2@v1.146.0 github.com/aws/aws-sdk-go-v2/service/internal/presigned-url@v1.10.10
github.com/aws/aws-sdk-go-v2/service/ec2@v1.146.0 github.com/aws/smithy-go@v1.19.0
github.com/aws/aws-sdk-go-v2/service/ec2@v1.146.0 github.com/google/go-cmp@v0.5.8
github.com/aws/aws-sdk-go-v2/service/ec2@v1.146.0 github.com/jmespath/go-jmespath@v0.4.0
github.com/aws/aws-sdk-go-v2/service/ecr@v1.24.7 github.com/aws/aws-sdk-go-v2@v1.24.1
github.com/aws/aws-sdk-go-v2/service/ecr@v1.24.7 github.com/aws/aws-sdk-go-v2/internal/configsources@v1.2.10
github.com/aws/aws-sdk-go-v2/service/ecr@v1.24.7 github.com/aws/aws-sdk-go-v2/internal/endpoints/v2@v2.5.10
github.com/aws/aws-sdk-go-v2/service/ecr@v1.24.7 github.com/aws/smithy-go@v1.19.0
github.com/aws/aws-sdk-go-v2/service/ecr@v1.24.7 github.com/google/go-cmp@v0.5.8
github.com/aws/aws-sdk-go-v2/service/ecr@v1.24.7 github.com/jmespath/go-jmespath@v0.4.0
github.com/aws/aws-sdk-go-v2/service/firehose@v1.24.0 github.com/aws/aws-sdk-go-v2@v1.24.1
github.com/aws/aws-sdk-go-v2/service/firehose@v1.24.0 github.com/aws/aws-sdk-go-v2/internal/configsources@v1.2.10
github.com/aws/aws-sdk-go-v2/service/firehose@v1.24.0 github.com/aws/aws-sdk-go-v2/internal/endpoints/v2@v2.5.10
github.com/aws/aws-sdk-go-v2/service/firehose@v1.24.0 github.com/aws/smithy-go@v1.19.0
github.com/aws/aws-sdk-go-v2/service/firehose@v1.24.0 github.com/google/go-cmp@v0.5.8
github.com/aws/aws-sdk-go-v2/service/kinesis@v1.24.7 github.com/aws/aws-sdk-go-v2@v1.24.1
github.com/aws/aws-sdk-go-v2/service/kinesis@v1.24.7 github.com/aws/aws-sdk-go-v2/aws/protocol/eventstream@v1.5.4
github.com/aws/aws-sdk-go-v2/service/kinesis@v1.24.7 github.com/aws/aws-sdk-go-v2/internal/configsources@v1.2.10
github.com/aws/aws-sdk-go-v2/service/kinesis@v1.24.7 github.com/aws/aws-sdk-go-v2/internal/endpoints/v2@v2.5.10
github.com/aws/aws-sdk-go-v2/service/kinesis@v1.24.7 github.com/aws/smithy-go@v1.19.0
github.com/aws/aws-sdk-go-v2/service/kinesis@v1.24.7 github.com/google/go-cmp@v0.5.8
github.com/aws/aws-sdk-go-v2/service/kinesis@v1.24.7 github.com/jmespath/go-jmespath@v0.4.0
github.com/aws/aws-sdk-go-v2/service/route53@v1.37.1 github.com/aws/aws-sdk-go-v2@v1.24.1
github.com/aws/aws-sdk-go-v2/service/route53@v1.37.1 github.com/aws/aws-sdk-go-v2/internal/configsources@v1.2.10
github.com/aws/aws-sdk-go-v2/service/route53@v1.37.1 github.com/aws/aws-sdk-go-v2/internal/endpoints/v2@v2.5.10
github.com/aws/aws-sdk-go-v2/service/route53@v1.37.1 github.com/aws/smithy-go@v1.19.0
github.com/aws/aws-sdk-go-v2/service/route53@v1.37.1 github.com/google/go-cmp@v0.5.8
github.com/aws/aws-sdk-go-v2/service/route53@v1.37.1 github.com/jmespath/go-jmespath@v0.4.0
github.com/aws/aws-sdk-go-v2/service/s3@v1.48.1 github.com/aws/aws-sdk-go-v2@v1.24.1
github.com/aws/aws-sdk-go-v2/service/s3@v1.48.1 github.com/aws/aws-sdk-go-v2/aws/protocol/eventstream@v1.5.4
github.com/aws/aws-sdk-go-v2/service/s3@v1.48.1 github.com/aws/aws-sdk-go-v2/internal/configsources@v1.2.10
github.com/aws/aws-sdk-go-v2/service/s3@v1.48.1 github.com/aws/aws-sdk-go-v2/internal/endpoints/v2@v2.5.10
github.com/aws/aws-sdk-go-v2/service/s3@v1.48.1 github.com/aws/aws-sdk-go-v2/internal/v4a@v1.2.10
github.com/aws/aws-sdk-go-v2/service/s3@v1.48.1 github.com/aws/aws-sdk-go-v2/service/internal/accept-encoding@v1.10.4
github.com/aws/aws-sdk-go-v2/service/s3@v1.48.1 github.com/aws/aws-sdk-go-v2/service/internal/checksum@v1.2.10
github.com/aws/aws-sdk-go-v2/service/s3@v1.48.1 github.com/aws/aws-sdk-go-v2/service/internal/presigned-url@v1.10.10
github.com/aws/aws-sdk-go-v2/service/s3@v1.48.1 github.com/aws/aws-sdk-go-v2/service/internal/s3shared@v1.16.10
github.com/aws/aws-sdk-go-v2/service/s3@v1.48.1 github.com/aws/smithy-go@v1.19.0
github.com/aws/aws-sdk-go-v2/service/s3@v1.48.1 github.com/google/go-cmp@v0.5.8
github.com/aws/aws-sdk-go-v2/service/secretsmanager@v1.26.2 github.com/aws/aws-sdk-go-v2@v1.24.1
github.com/aws/aws-sdk-go-v2/service/secretsmanager@v1.26.2 github.com/aws/aws-sdk-go-v2/internal/configsources@v1.2.10
github.com/aws/aws-sdk-go-v2/service/secretsmanager@v1.26.2 github.com/aws/aws-sdk-go-v2/internal/endpoints/v2@v2.5.10
github.com/aws/aws-sdk-go-v2/service/secretsmanager@v1.26.2 github.com/aws/smithy-go@v1.19.0
github.com/aws/aws-sdk-go-v2/service/secretsmanager@v1.26.2 github.com/google/go-cmp@v0.5.8
github.com/aws/aws-sdk-go-v2/service/sts@v1.26.7 github.com/aws/aws-sdk-go-v2@v1.24.1
github.com/aws/aws-sdk-go-v2/service/sts@v1.26.7 github.com/aws/aws-sdk-go-v2/internal/configsources@v1.2.10
github.com/aws/aws-sdk-go-v2/service/sts@v1.26.7 github.com/aws/aws-sdk-go-v2/internal/endpoints/v2@v2.5.10
github.com/aws/aws-sdk-go-v2/service/sts@v1.26.7 github.com/aws/aws-sdk-go-v2/service/internal/accept-encoding@v1.10.4
github.com/aws/aws-sdk-go-v2/service/sts@v1.26.7 github.com/aws/aws-sdk-go-v2/service/internal/presigned-url@v1.10.10
github.com/aws/aws-sdk-go-v2/service/sts@v1.26.7 github.com/aws/smithy-go@v1.19.0
github.com/aws/aws-sdk-go-v2/service/sts@v1.26.7 github.com/google/go-cmp@v0.5.8
github.com/aws/aws-sdk-go-v2/aws/protocol/eventstream@v1.5.4 github.com/aws/smithy-go@v1.19.0
github.com/aws/aws-sdk-go-v2/feature/dynamodb/attributevalue@v1.12.16 github.com/aws/aws-sdk-go-v2@v1.24.1
github.com/aws/aws-sdk-go-v2/feature/dynamodb/attributevalue@v1.12.16 github.com/aws/aws-sdk-go-v2/service/dynamodb@v1.27.0
github.com/aws/aws-sdk-go-v2/feature/dynamodb/attributevalue@v1.12.16 github.com/aws/aws-sdk-go-v2/service/dynamodbstreams@v1.18.7
github.com/aws/aws-sdk-go-v2/feature/dynamodb/attributevalue@v1.12.16 github.com/aws/smithy-go@v1.19.0
github.com/aws/aws-sdk-go-v2/feature/dynamodb/attributevalue@v1.12.16 github.com/google/go-cmp@v0.5.8
github.com/aws/aws-sdk-go-v2/feature/dynamodb/attributevalue@v1.12.16 github.com/jmespath/go-jmespath@v0.4.0
github.com/aws/aws-sdk-go-v2/feature/ec2/imds@v1.14.11 github.com/aws/aws-sdk-go-v2@v1.24.1
github.com/aws/aws-sdk-go-v2/feature/ec2/imds@v1.14.11 github.com/aws/smithy-go@v1.19.0
github.com/aws/aws-sdk-go-v2/feature/ec2/imds@v1.14.11 github.com/google/go-cmp@v0.5.8
github.com/aws/aws-sdk-go-v2/internal/configsources@v1.2.10 github.com/aws/aws-sdk-go-v2@v1.24.1
github.com/aws/aws-sdk-go-v2/internal/configsources@v1.2.10 github.com/aws/smithy-go@v1.19.0
github.com/aws/aws-sdk-go-v2/internal/endpoints/v2@v2.5.10 github.com/aws/aws-sdk-go-v2@v1.24.1
github.com/aws/aws-sdk-go-v2/internal/endpoints/v2@v2.5.10 github.com/aws/smithy-go@v1.19.0
github.com/aws/aws-sdk-go-v2/internal/endpoints/v2@v2.5.10 github.com/google/go-cmp@v0.5.8
github.com/aws/aws-sdk-go-v2/internal/v4a@v1.2.10 github.com/aws/aws-sdk-go-v2@v1.24.1
github.com/aws/aws-sdk-go-v2/internal/v4a@v1.2.10 github.com/aws/smithy-go@v1.19.0
github.com/aws/aws-sdk-go-v2/internal/v4a@v1.2.10 github.com/google/go-cmp@v0.5.8
github.com/aws/aws-sdk-go-v2/service/dynamodb@v1.27.0 github.com/aws/aws-sdk-go-v2@v1.24.1
github.com/aws/aws-sdk-go-v2/service/dynamodb@v1.27.0 github.com/aws/aws-sdk-go-v2/internal/configsources@v1.2.10
github.com/aws/aws-sdk-go-v2/service/dynamodb@v1.27.0 github.com/aws/aws-sdk-go-v2/internal/endpoints/v2@v2.5.10
github.com/aws/aws-sdk-go-v2/service/dynamodb@v1.27.0 github.com/aws/aws-sdk-go-v2/service/internal/accept-encoding@v1.10.4
github.com/aws/aws-sdk-go-v2/service/dynamodb@v1.27.0 github.com/aws/aws-sdk-go-v2/service/internal/endpoint-discovery@v1.8.11
github.com/aws/aws-sdk-go-v2/service/dynamodb@v1.27.0 github.com/aws/smithy-go@v1.19.0
github.com/aws/aws-sdk-go-v2/service/dynamodb@v1.27.0 github.com/google/go-cmp@v0.5.8
github.com/aws/aws-sdk-go-v2/service/dynamodb@v1.27.0 github.com/jmespath/go-jmespath@v0.4.0
github.com/aws/aws-sdk-go-v2/service/dynamodbstreams@v1.18.7 github.com/aws/aws-sdk-go-v2@v1.24.1
github.com/aws/aws-sdk-go-v2/service/dynamodbstreams@v1.18.7 github.com/aws/aws-sdk-go-v2/internal/configsources@v1.2.10
github.com/aws/aws-sdk-go-v2/service/dynamodbstreams@v1.18.7 github.com/aws/aws-sdk-go-v2/internal/endpoints/v2@v2.5.10
github.com/aws/aws-sdk-go-v2/service/dynamodbstreams@v1.18.7 github.com/aws/smithy-go@v1.19.0
github.com/aws/aws-sdk-go-v2/service/dynamodbstreams@v1.18.7 github.com/google/go-cmp@v0.5.8
github.com/aws/aws-sdk-go-v2/service/internal/accept-encoding@v1.10.4 github.com/aws/smithy-go@v1.19.0
github.com/aws/aws-sdk-go-v2/service/internal/checksum@v1.2.10 github.com/aws/aws-sdk-go-v2@v1.24.1
github.com/aws/aws-sdk-go-v2/service/internal/checksum@v1.2.10 github.com/aws/aws-sdk-go-v2/service/internal/presigned-url@v1.10.10
github.com/aws/aws-sdk-go-v2/service/internal/checksum@v1.2.10 github.com/aws/smithy-go@v1.19.0
github.com/aws/aws-sdk-go-v2/service/internal/checksum@v1.2.10 github.com/google/go-cmp@v0.5.8
github.com/aws/aws-sdk-go-v2/service/internal/endpoint-discovery@v1.8.11 github.com/aws/aws-sdk-go-v2@v1.24.1
github.com/aws/aws-sdk-go-v2/service/internal/endpoint-discovery@v1.8.11 github.com/aws/smithy-go@v1.19.0
github.com/aws/aws-sdk-go-v2/service/internal/presigned-url@v1.10.10 github.com/aws/aws-sdk-go-v2@v1.24.1
github.com/aws/aws-sdk-go-v2/service/internal/presigned-url@v1.10.10 github.com/aws/smithy-go@v1.19.0
github.com/aws/aws-sdk-go-v2/service/internal/presigned-url@v1.10.10 github.com/google/go-cmp@v0.5.8
github.com/aws/aws-sdk-go-v2/service/internal/s3shared@v1.16.10 github.com/aws/aws-sdk-go-v2@v1.24.1
github.com/aws/aws-sdk-go-v2/service/internal/s3shared@v1.16.10 github.com/aws/smithy-go@v1.19.0
github.com/aws/aws-sdk-go-v2/service/sso@v1.18.7 github.com/aws/aws-sdk-go-v2@v1.24.1
github.com/aws/aws-sdk-go-v2/service/sso@v1.18.7 github.com/aws/aws-sdk-go-v2/internal/configsources@v1.2.10
github.com/aws/aws-sdk-go-v2/service/sso@v1.18.7 github.com/aws/aws-sdk-go-v2/internal/endpoints/v2@v2.5.10
github.com/aws/aws-sdk-go-v2/service/sso@v1.18.7 github.com/aws/smithy-go@v1.19.0
github.com/aws/aws-sdk-go-v2/service/sso@v1.18.7 github.com/google/go-cmp@v0.5.8
github.com/aws/aws-sdk-go-v2/service/ssooidc@v1.21.7 github.com/aws/aws-sdk-go-v2@v1.24.1
github.com/aws/aws-sdk-go-v2/service/ssooidc@v1.21.7 github.com/aws/aws-sdk-go-v2/internal/configsources@v1.2.10
github.com/aws/aws-sdk-go-v2/service/ssooidc@v1.21.7 github.com/aws/aws-sdk-go-v2/internal/endpoints/v2@v2.5.10
github.com/aws/aws-sdk-go-v2/service/ssooidc@v1.21.7 github.com/aws/smithy-go@v1.19.0
github.com/aws/aws-sdk-go-v2/service/ssooidc@v1.21.7 github.com/google/go-cmp@v0.5.8
github.com/aws/smithy-go@v1.19.0 github.com/google/go-cmp@v0.5.8
github.com/aws/aws-sdk-go@v1.19.11 github.com/jmespath/go-jmespath@v0.0.0-20180206201540-c2b33e8439af
github.com/aws/aws-sdk-go@v1.30.27 github.com/go-sql-driver/mysql@v1.5.0
github.com/aws/aws-sdk-go@v1.30.27 github.com/jmespath/go-jmespath@v0.3.0
github.com/aws/aws-sdk-go@v1.30.27 github.com/pkg/errors@v0.9.1
github.com/aws/aws-sdk-go@v1.30.27 golang.org/x/net@v0.0.0-20200202094626-16171245cfb2
github.com/aws/aws-sdk-go@v1.37.32 github.com/jmespath/go-jmespath@v0.4.0
github.com/aws/aws-sdk-go@v1.37.32 github.com/pkg/errors@v0.9.1
github.com/aws/aws-sdk-go@v1.37.32 golang.org/x/net@v0.0.0-20201110031124-69a78807bb2b
github.com/aws/aws-sdk-go@v1.30.19 github.com/go-sql-driver/mysql@v1.5.0
github.com/aws/aws-sdk-go@v1.30.19 github.com/jmespath/go-jmespath@v0.3.0
github.com/aws/aws-sdk-go@v1.30.19 github.com/pkg/errors@v0.9.1
github.com/aws/aws-sdk-go@v1.30.19 golang.org/x/net@v0.0.0-20200202094626-16171245cfb2
github.com/aws/aws-lambda-go@v1.13.3 github.com/stretchr/testify@v1.4.0
github.com/aws/aws-lambda-go@v1.13.3 github.com/urfave/cli@v1.22.1
github.com/aws/aws-sdk-go@v1.27.0 github.com/jmespath/go-jmespath@v0.0.0-20180206201540-c2b33e8439af
github.com/aws/aws-sdk-go-v2@v0.18.0 github.com/davecgh/go-spew@v1.1.1
github.com/aws/aws-sdk-go-v2@v0.18.0 github.com/go-sql-driver/mysql@v1.4.0
github.com/aws/aws-sdk-go-v2@v0.18.0 github.com/google/go-cmp@v0.3.1
github.com/aws/aws-sdk-go-v2@v0.18.0 github.com/jmespath/go-jmespath@v0.0.0-20180206201540-c2b33e8439af
github.com/aws/aws-sdk-go-v2@v0.18.0 github.com/pmezard/go-difflib@v1.0.0
github.com/aws/aws-sdk-go-v2@v0.18.0 github.com/stretchr/testify@v1.2.2
github.com/aws/aws-sdk-go-v2@v0.18.0 golang.org/x/net@v0.0.0-20181201002055-351d144fa1fc
github.com/aws/aws-sdk-go-v2@v0.18.0 google.golang.org/appengine@v1.2.0

### Compiler and Version used

go version go1.21.6 linux/amd64

### Operating System and version

ubuntu

**Comment by lucix-aws** [comment]
@hjkatz --

This is ultimately a fault in server-side behavior. The IAM state change associated with retrieving credentials (via AssumeRole or whatever else) do not propagate immediately after credentials are returned.

**Comment by RanVaknin** [comment]
Hi @hjkatz ,

Just to chime in as well. This is not an SDK issue and is not unique to SSO. What you are experiencing is a propagation delay that is solved with a retry. This is not an issue with the SDK itself but just a nature of a distributed system.

Thanks,
Ran~

**Comment by hjkatz** [comment]
> Hi @hjkatz ,
> 
> Just to chime in as well. This is not an SDK issue and is not unique to SSO. What you are experiencing is a propagation delay that is solved with a retry. This is not an issue with the SDK itself but just a nature of a distributed system.
> 
> Thanks,
> Ran~

@RanVaknin Kindly review the linked discussion. Retrying 5 times with a gradual bockoff from 500ms to 2s still reproduces the issue.

I understand that a distributed system will require time to propogate the new token. So I'll start with my goal instead: I want to verify that the SSO session via the SDK is valid. How can I test that reliably?

**Comment by hjkatz** [comment]
@RanVaknin I realize that the discussion doest have as clear of an example as I'm suggesting. I'll upload a clear example later today.

**Comment by hjkatz** [comment]
```go
/ Wrapper for sts.GetCallerIdentity() that supports retries
//
// Use the context to set a maximum time that retries can be performed.
// When the context is canceled then the final response will be returned.
//
// See: https://github.com/aws/aws-sdk-go-v2/discussions/2093#discussioncomment-8455830
func StsGetCallerIdentity(ctx context.Context, client *sts.Client) (result *sts.GetCallerIdentityOutput, err error) {
    if client == nil {
        return nil, errs.New("cannot call StsGetCallerIdentity with nil client!")
    }

    attempt := 0
    retries := 5
    // internal lib that implements a backoff between start -> end, without jitter (false)
    backoff := reliable.NewBackoff(100*time.Millisecond, 1*time.Second, false)

    for attempt < retries {
        attempt++

        select {
        case <-ctx.Done():
            // ran out of time, return whatever we got last
            return
        default:
            // continue below
        }

        result, err = client.GetCallerIdentity(ctx, &sts.GetCallerIdentityInput{})
        if err == nil {
            // success
            return // final values
        }

        // See: https://github.com/aws/aws-sdk-go-v2/discussions/2093
        if !strings.Contains(err.Error(), "api error InvalidClientTokenId") {
            // non-retryable error
            return // last result + error
        }

        // error, backoff and try again
        be := backoff.Wait(ctx) // Wait() == time.Sleep(backoff.NextDuration())
        if be != nil {
            // context canceled
            return // last values
        }
    }

    return // last values
}
```

I hope this helps reproduce what we're seeing.

I'm also hopeful we can find a way via the SDK to verify the session credentials returned by the SSO credentials provider are valid. Some thoughts to get at this information:
* sts get-caller-identity (as above)
* reading a profile from `~/.aws/config` that depends on `sso_session = my-session` and see if it errors or not (relying on internal `aws` CLI's boto implementation)
* reading refresh token from the SSO provider directly? (not yet tried) see: https://pkg.go.dev/github.com/aws/aws-sdk-go-v2/credentials/ssocreds
* Some new implementation in `ssocreds` that provides this ability

Happy to discuss alternatives too!

**Comment by lucix-aws** [comment]
@hjkatz On average, how long does it take for `GetCallerIdentity` to recognize the token?

This _really_ seems like it should be something that sts models as `@waitable`. That would enable automatic code generation of a waiter API that polls `GetCallerIdentity` until a success response is returned.

In case you're not familiar with waiters -- https://aws.github.io/aws-sdk-go-v2/docs/making-requests/#using-waiters.

The most ubiquitous example in my mind would be `dynamodb.TableExistsWaiter`, since ddb table creates are technically asynchronous.

tl;dr waiting for async state changes is a problem we've solved at large, if I understand correctly it's just a question of pushing for sts to add some additional modeling to solve this specific case

**Comment by hjkatz** [comment]
> Followup - what is the delay between you provisioning the token (looks like through sso) and first calling `GetCallerIdentity`?

It's on the order of milliseconds.

For context we have a shared developer CLI that everyone uses for various commands/tools/utilities. In that CLI we annotate some commands as needing SSO to work correctly. Many of our commands are annotated and interact with AWS in some required way.

Our goal was to warn the user that they have not started their SSO session for the running command. To do this we need to check if the session is valid (not sure how to do this), and we came up with generating a token then trying to see if it works.

**Comment by hjkatz** [comment]
I gotcha. I was writing up a test case to get some real data.

```
Summary:
1ms: 143323
100us: 856294
10ms: 374
100ms: 9
```

Here's the summary of 1 million attempts. It's looking _much_ better today than when I originally opened the ticket. For whatever reason I'm not seeing anything take longer than ~100ms but in the past I would feel the delay more in the ~1-2s range, so maybe something's improved. (It could be my network as I'm at my parents' place atm)

How about I run this test again daily and get back to you with more data?

**Comment by hjkatz** [comment]
Today's summary also seems fine:

```
Summary:
10ms: 359
1ms: 141759
100us: 857875
100ms: 7
```

I'm suspecting that the way I'm using a backoff after getting the credentials should be a naive loop with `context.WithTimeout()` rather than some arbitrary number of `attempts`. Like maybe it takes 10ms, but that's 20 attempts, yet I'm exiting after 5 attempts. I'll keep tracking this.

**Comment by hjkatz** [comment]
Today seems the same too.

```
Summary:
10ms: 455
1ms: 160674
100us: 838859
100ms: 12
```

I'm going to add additional logging into our CLI and see if I can reproduce any inconsistency today, but my suspicion is that 5 attempts isn't enough and we just need to try more times.

**Comment by lucix-aws** [comment]
My recommendation in general there would not be to fix the number of attempts and instead write your "waiting" construct to accept a timeout.

This is how modeled waiters are written (well, generated, but obviously we wrote the code to do that). The waiter then just retries "infinitely" (with an increasing backoff) until it either hits the success case or exceeds the caller-provided deadline.
